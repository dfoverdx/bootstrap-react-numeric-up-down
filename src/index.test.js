import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';
import renderer from 'react-test-renderer';

jest.mock('./fortawesome');
Enzyme.configure({adapter: new Adapter()});

const NumericUpDown = require('./index').default;

it('honors min and max when clicking buttons', () => {
    let nud,
        onChange = sinon.spy(v => nud.setProps({ value: v }));

    nud = shallow(
        <NumericUpDown type="number" min={-1} max={1} value={0} onChange={onChange} />
    );

    expect(nud.props().value).toBe(0);
    
    nud.find('.minus-icon').simulate('click');
    expect(onChange.callCount).toBe(1);
    expect(onChange.getCall(0).args[0]).toBe(-1);
    expect(nud.props().value).toBe(-1);

    onChange.resetHistory();

    expect(nud.find('.minus-icon').hasClass('disabled')).toBe(true);
    nud.find('.minus-icon').simulate('click');
    expect(onChange.notCalled).toBe(true);
    expect(nud.props().value).toBe(-1);

    nud.setProps({ value: 0 });
    onChange.resetHistory();

    nud.find('.plus-icon').simulate('click');
    expect(onChange.callCount).toBe(1);
    expect(onChange.getCall(0).args[0]).toBe(1);
    expect(nud.props().value).toBe(1);
    
    onChange.resetHistory();
    nud.setProps({ value: 1 });
    nud.find('.plus-icon').simulate('click');
    expect(onChange.notCalled).toBe(true);
    expect(nud.props().value).toBe(1);
});

it('restores the target value when the target fits the range', () => {
    let nud,
        onChange = sinon.spy(v => nud.setProps({ value: v }));

    nud = shallow(
        <NumericUpDown type="number" min={0} max={3} value={0} onChange={onChange} />
    );

    expect(nud.props().value).toBe(0);
    nud.setProps({ min: 1 });
    expect(onChange.callCount).toBe(1);
    expect(onChange.getCall(0).args[0]).toBe(1);
    expect(nud.props().value).toBe(1);
    expect(nud.state().targetValue).toBe(0);

    onChange.resetHistory();
    nud.setProps({ min: 0 });
    expect(onChange.callCount).toBe(1);
    expect(onChange.getCall(0).args[0]).toBe(0);
    expect(nud.state().targetValue).toBeNull();

    nud.setProps({ min: 4 });
    onChange.resetHistory();
    expect(nud.props().value).toBe(3);
    expect(nud.find('input').props().disabled).toBe(true);
    expect(nud.state().targetValue).toBe(0);

    onChange.resetHistory();
    nud.setProps({ max: 5 });
    expect(onChange.callCount).toBe(1);
    expect(onChange.getCall(0).args[0]).toBe(4);
    expect(nud.props().value).toBe(4);
    expect(nud.state().targetValue).toBe(0);

    onChange.resetHistory();
    nud.setProps({ min: -Infinity });
    expect(onChange.callCount).toBe(1);
    expect(onChange.getCall(0).args[0]).toBe(0);
    expect(nud.props().value).toBe(0);
    expect(nud.state().targetValue).toBeNull();
});

it('restores a value if original props have empty range', () => {
    let nud,
        onChange = sinon.spy(v => nud.setProps({ value: v }));

    nud = shallow(
        <NumericUpDown type="number" min={1} max={0} value={1} onChange={onChange} />
    );

    expect(nud.props().value).toBe(1);
    expect(nud.state().targetValue).toBe(1);

    nud.setProps({ max: 3 });
    expect(onChange.notCalled).toBe(true);
    expect(nud.props().value).toBe(1);
    expect(nud.state().targetValue).toBeNull();
});

describe('icon props', () => {
    const fortawesome = require('./fortawesome');

    beforeEach(() => {
        jest.resetModules();
        fortawesome.__setReactFontawesomeInstalled(true);
    });

    // it(`uses '+' and '-' characters by default when react-fontawesome is not installed`, () => {
    //     fortawesome.__setReactFontawesomeInstalled(false);
    //     const NUD = require('./index').default;
    //     let nud = renderer.create(<NUD value={0} />),
    //         tree = nud.toJSON();
    //     expect(tree).toMatchSnapshot();
    // });
    
    it(`uses faPlusCircle and faMinusCircle by default when react-fontawesome and free-solid-svg-icons are installed`, () => {
        const NUD = require('./index').default;
        let nud = renderer.create(<NUD value={0} />),
            tree = nud.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test(`uses specified fontawesome icons when react-fontawesome and free-solid-svg-icons are installed`, () => {
        const NUD = require('./index').default,
            { faUserMinus, faUserPlus } = require('@fortawesome/free-solid-svg-icons');

        let nud = renderer.create(<NUD value={0} minusIcon={faUserMinus} plusIcon={faUserPlus} />),
            tree = nud.toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it(`handles font-awesome strings for icons when specified in properties`, () => {
        const NUD = require('./index').default,
            { library } = require('@fortawesome/fontawesome-svg-core'),
            { fas } = require('@fortawesome/free-solid-svg-icons');
        library.add(fas);

        let nud = renderer.create(<NUD value={0} minusIcon="user-minus" plusIcon="user-plus" />),
            tree = nud.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
