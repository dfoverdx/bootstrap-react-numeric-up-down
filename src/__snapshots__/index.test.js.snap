// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon props handles font-awesome strings for icons when specified in properties 1`] = `
<div
  className="jsx-3892955365 d-flex flex-nowrap flex-row align-items-center numeric-up-down"
  max={Infinity}
  min={0}
  step={1}
  value={0}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-user-minus fa-w-20 minus-icon disabled jsx-1158657347"
    data-icon="user-minus"
    data-prefix="fas"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M624 208H432c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h192c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <input
    className="jsx-3892955365 form-control mx-2"
    max={Infinity}
    min={0}
    step={1}
    type="number"
    value={0}
  />
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-user-plus fa-w-20 plus-icon jsx-1158657347"
    data-icon="user-plus"
    data-prefix="fas"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M624 208h-64v-64c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v64h-64c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h64v64c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-64h64c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;

exports[`icon props uses faPlusCircle and faMinusCircle by default when react-fontawesome and free-solid-svg-icons are installed 1`] = `
<div
  className="jsx-3892955365 d-flex flex-nowrap flex-row align-items-center numeric-up-down"
  max={Infinity}
  min={0}
  step={1}
  value={0}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-minus-circle fa-w-16 minus-icon disabled jsx-1158657347"
    data-icon="minus-circle"
    data-prefix="fas"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <input
    className="jsx-3892955365 form-control mx-2"
    max={Infinity}
    min={0}
    step={1}
    type="number"
    value={0}
  />
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-plus-circle fa-w-16 plus-icon jsx-1158657347"
    data-icon="plus-circle"
    data-prefix="fas"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;

exports[`icon props uses specified fontawesome icons when react-fontawesome and free-solid-svg-icons are installed 1`] = `
<div
  className="jsx-3892955365 d-flex flex-nowrap flex-row align-items-center numeric-up-down"
  max={Infinity}
  min={0}
  step={1}
  value={0}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-user-minus fa-w-20 minus-icon disabled jsx-1158657347"
    data-icon="user-minus"
    data-prefix="fas"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M624 208H432c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h192c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <input
    className="jsx-3892955365 form-control mx-2"
    max={Infinity}
    min={0}
    step={1}
    type="number"
    value={0}
  />
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-user-plus fa-w-20 plus-icon jsx-1158657347"
    data-icon="user-plus"
    data-prefix="fas"
    onClick={[Function]}
    role="img"
    style={Object {}}
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M624 208h-64v-64c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v64h-64c-8.8 0-16 7.2-16 16v32c0 8.8 7.2 16 16 16h64v64c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-64h64c8.8 0 16-7.2 16-16v-32c0-8.8-7.2-16-16-16zm-400 48c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;
